### DISCUSSION TITLE
Opening a websocket seems to pause existing Background Task

### DISCUSSION THREAD

**Discussion by dbatten5** [fastapi/fastapi#6741] [Questions]
Most likely this is an issue with my current setup, I'm quite new to async python so I'm sure I'm missing some conceptual things.

My use case is:
- Kick off a background task that involves adding database records
- After each db record addition, the task updates a `progress` attribute in a `Job` model and sleeps for a bit to avoid getting rate limited by an external service
- While this task is processing, stream the `progress` back to the client via a websocket

### Example

This is just to illustrate the problem, won't work as is

```Python
import time
import asyncio

from fastapi import FastAPI, BackgroundTasks, WebSocket
from pydantic import BaseModel


class Job(BaseModel):
    id: int
    progress: float = 0.0


app = FastAPI()


def long_running_task(job: Job):
    for x in range(10):
        db_job = # fetch job from database
        # add a database entry
        db_job.progress = (x / 10) * 100
        # commit to database
        time.sleep(20) # needed to avoid getting rate limited by an external service


@app.post('/tasks/', response_model=Job)
def create_task(background_tasks: BackgroundTasks):
    job = Job()
    # commit job to database
    background_tasks.add_task(long_running_task, job)
    return job


@app.websocket("/job/{job_id}/ws")
async def websocket_endpoint(websocket: WebSocket, job_id: str):
    await websocket.accept()
    while True:
        await asyncio.sleep(2)
        job = # fetch job from database
        await websocket.send_json(jsonable_encoder(job))
```

### Description

The background task is kicked off fine and the database entries are added as expected. However when I open the websocket the background task seems to pause and remains paused until the websocket is closed. At that point the background task starts again and carries on as normal. How can I open the websocket without causing the task to stop execution?

I read somewhere about using `asyncio.ensure_future` but I'm not quite sure how to use it in this context. Also quite unsure about my current setup and whether the task should be async. When I switch it to `async def` (and `await asyncio.sleep(20)`) though the post request takes a long time to return a response.

### Environment

* OS: macOS
* FastAPI Version: 0.63.0
* Python version: 3.9.2

Thanks in advance for any help

**UPDATE**
After adding some logging to my job it turns it is was actually being run, it's just that the job I was fetching from the database before sending a message through the websocket wasn't updated with the latest progress. After running a `db.refresh(job)` it worked.

**Comment by coder3112** [comment]
> UPDATE
After adding some logging to my job it turns it is was actually being run, it's just that the job I was fetching from the database before sending a message through the websocket wasn't updated with the latest progress. After running a db.refresh(job) it worked.

Please close the issue if it is solved.

**Comment by dbatten5** [comment]
@coder3122 will do - was leaving it open in case anyone had any advice on using async in this scenario

**Comment by tiangolo** [comment]
Thanks for the help here @coder3112 ! 👏  🙇 

Thanks for reporting back and closing the issue @dbatten5  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
